src/net/neighbors/walk.rs

```diff
@@ -0,0 +1,2342 @@
+            }
+        };
+
+        let mut neighbor_addrs = match message.payload {
+            StacksMessageType::Neighbors(ref data) => {
+                debug!(
+                    "{:?}: Got Neighbors from {:?}: {:?}",
+                    network.get_local_peer(),
+                    &self.cur_neighbor.addr,
+                    data.neighbors
+                );
+                let neighbors = NeighborWalk::filter_sensible_neighbors(data.neighbors.clone());
+                let (mut found, to_resolve) = NeighborWalk::lookup_stale_neighbors(
+                    network.peerdb.conn(),
+                    message.preamble.network_id,
+                    block_height,
+                    &neighbors,
+                )?;
+
+                // add neighbors we already know about to the frontier of `cur_neighbor`
+                for (_naddr, neighbor) in found.drain() {
+                    self.new_frontier
+                        .insert(neighbor.addr.clone(), neighbor.clone());
+                    self.frontier
+                        .insert(neighbor.addr.clone(), neighbor.clone());
+                }
+
+                to_resolve
+            }
+            StacksMessageType::Nack(ref data) => {
+                debug!(
+                    "{:?}: Neighbor {:?} NACK'ed GetNeighbors with code {:?}",
+                    network.get_local_peer(),
+                    &self.cur_neighbor.addr,
+                    data.error_code
+                );
+                self.add_broken(network, &self.cur_neighbor.addr.clone());
+                return Err(net_error::ConnectionBroken);
+            }
+            _ => {
+                // invalid message
+                debug!(
+                    "{:?}: Got out-of-sequence message from {:?}",
+                    network.get_local_peer(),
+                    &self.cur_neighbor.addr
+                );
+                self.add_broken(network, &self.cur_neighbor.addr.clone());
+                return Err(net_error::InvalidMessage);
+            }
+        };
+
+        // proceed to handshake with them.
+        // If this is an inbound neighbor, then try also to handshake with its advertized
+        // IP address.
+        if !self.walk_outbound {
+            test_debug!("{:?}: will try to handshake with inbound neighbor {:?}'s advertized address {:?} as well", network.get_local_peer(), &self.cur_neighbor.addr, &self.neighbor_from_handshake);
+            let cur_neighbor_pubkey_hash =
+                Hash160::from_node_public_key(&self.cur_neighbor.public_key);
+            neighbor_addrs.push(NeighborAddress::from_neighbor_key(
+                self.neighbor_from_handshake.clone(),
+                cur_neighbor_pubkey_hash,
+            ));
+        }
+
+        // prune the list to a reasonable size in case cur_neighbor gave us too many for our
+        // configuration
+        if neighbor_addrs.len() as u64 > network.get_connection_opts().max_neighbors_of_neighbor {
```

Comment:
I'm not sure how whether or not the peer is inbound is relevant here?  This is to prevent a denial of service attack.  Suppose the remote node gave me a 16MB list of neighbors, each of which took a long time to respond.

Reply:
Immediately before this pruning, the node checks if it is doing an inbound walk, and adds the cur_neighbor to the to_resolve list:

```
        if !self.walk_outbound {
            test_debug!("{:?}: will try to handshake with inbound neighbor {:?}'s advertized address {:?} as well", network.get_local_peer(), &self.cur_neighbor.addr, &self.neighbor_from_handshake);
            let cur_neighbor_pubkey_hash =
                Hash160::from_node_public_key(&self.cur_neighbor.public_key);
            neighbor_addrs.push(NeighborAddress::from_neighbor_key(
                self.neighbor_from_handshake.clone(),
                cur_neighbor_pubkey_hash,
            ));
        }
```

Pruning _after_ that means that a neighbor could return a maximally valid list, then this node would add `cur_neighbor` to the "to resolve" list, and prune the `cur_neighbor`.

Reply:
Ah, yes. You're right; pruning before would be better here.  Will change.